pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -914,6 +1086,87 @@ func (r *ReconcileClusterPool) setAvailableCapacityCondition(pool *hivev1.Cluste
 	return nil
 }
 
+func (r *ReconcileClusterPool) setInventoryValidCondition(pool *hivev1.ClusterPool, add bool, cdcName string, state string, logger log.FieldLogger) error {
```

Comment:
I stared at this function for a while. It feels really complicated. I understand why: one individual call is supposed to be adding/changing/removing the state for a given CDC from a message that already has an arbitrary number of CDCs' states in it.

I dig making the message a JSON string, though I had to get used to the idea.

But here's an idea to make this function a lot simpler: What if we reverse the map? So the keys are the CDC names and the values are the states. That way the function becomes roughly:
```go
currentCondition := controllerutils.FindClusterPoolCondition(pool.Status.Conditions, hivev1.ClusterPoolInventoryValidCondition)

// Decode the current message
curMap := map[string]string{}
if err := json.Unmarshal([]byte(currentCondition.Message, &curMap); err != nil {
	// curMap unaffected.
	// log a warning that we're using an empty map / clearing the message
}

// Replace the entry for this CDC -- but omit if it's valid
delete(curMap, cdcName)
if state != cdcStateValid { // NB: assuming we make a typedef/enum for the states
	curMap[cdcName] = state
}

// Default condition settings to "valid"
status := corev1.ConditionTrue
reason := InventoryValidReasonValid  // enum/typedef rather than hardcoded string
message := "All inventory entries valid"
// If there are any entries in the map, at least one CDC is bad
if len(curMap) != 0 {
	status = corev1.ConditionFalse
	reason = InventoryValidReasonInvalid  // ditto
	messageByte, err := json.Marshal(curMap)
	message = string(messageByte)
}

// Do the status update (as you have it already)
// ...
```

This does make the message potentially longer (any time >1 CDC is in the same state), but I think that's acceptable for how much it simplifies the code. If it really bugs you though, we could reverse it after/before we Unmarshal/Marshal it with a simple loop.